### ISSUE TITLE
specific values for scale parameter can break dims

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#4861]
## 🐛 Bug
haven't dug deeper yet, but reposting here from
https://forum.image.sc/t/image-layer-in-napari-showing-the-wrong-dimension-size-one-plane-is-missing/69939/10

```python
d = np.random.rand(5, 170, 240)
viewer = napari.view_image(d, scale=[3.533, 1.0, 1.0])
```

slider only has 4 positions (not 5) ... where a slightly different first scale value works fine

**Comment by brisvag** [comment]
Found the problem. In short:

```py
(17.665-0.0) // 3.533 == 4.0  # what we do in `Dims.nsteps`
(17.665-0.0) / 3.533 == 5.0  # what apparently we should do
```

Some floating point error that's so small it's not even displayed, I suppose. The result of the above expression is passed to `int()`, so it's floored anyways... so I supposed we don't need the floor division. But this makes me wonder if the opposite case can happen as well? (i.e: we "spill" into the next integer)

**Comment by manerotoni** [comment]
Hello ,
we would like to reopen this issue as we believe that the problem is still present. 

Please see discussion in https://github.com/NEUBIAS/training-resources/issues/563.

For certain images and scaling settings, Napari does not show some Z-slices (typically the last one). There is an example code in https://neubias.github.io/training-resources/multidimensional_image_basics/index.html#xyzc (`drop-down menu > skimage naperi`). The loading function OpenIJTiff can be found here https://neubias.github.io/training-resources/functions/OpenIJTIFF.py

The image has 41 slices, when loaded with certain value of scales (certain values) it shows in the slider 0-39 slices. The last slices is just not displayed.  We observed the issue on several installations.

These are the Conda napari versions installed
```
napari                    0.4.17          pyh275ddea_0_pyqt    conda-forge
napari-console            0.0.7              pyhd8ed1ab_0    conda-forge
napari-plot-profile       0.2.2                    pypi_0    pypi
napari-plugin-engine      0.2.0              pyhd8ed1ab_2    conda-forge
napari-skimage-regionprops 0.10.0                   pypi_0    pypi
napari-svg                0.1.6              pyhd8ed1ab_1    conda-forge
napari-tools-menu         0.1.19                   pypi_0    pypi
napari-workflows          0.2.8                    pypi_0    pypi
```

**Comment by jni** [comment]
@manerotoni Could you please check the version in main? There have been some changes with the way that we compute extents in the last few weeks actually.

Also if you have a specific stack that we can test with that would be very useful!

**Comment by psobolewskiPhD** [comment]
@manerotoni 
Hi, thanks for reporting. 
I tried:
```
import numpy as np
import napari
dummy = np.random.rand(41, 2, 297, 284)
viewer = napari.view_image(dummy, scale=[0.1000000, 1, 0.0222057, 0.0222057])
```
On 0.4.17 i get a slider from 0 to 39, so it reproduces the bug.

However, in my dev env with main napari, I get 0-40, so it appears that it has been fixed or the behavior is somehow different.
I'll look into it more.

Also, can you provide full `napari --info`? As it's related to floats it could be platform dependent? I'm on macOS arm64.

**Comment by psobolewskiPhD** [comment]
Ok, git bisect points to this PR as fixing the issue:
https://github.com/napari/napari/pull/5751

@brisvag @Czaki that PR is tagged for 0.5.0 but this seems like something people can easily and frequently hit in the wild: with the above snippet it seems to occur anytime the number of slices is odd.
Should we reconsider?

**Comment by brisvag** [comment]
Just for clarity, this is actually a *different* bug (before it was a floating point math precision issue, now it's a "excluded vs included upper limit" issue.

#5751 is marked as 0.5.0 because it's actually a followup on the (quite significant) changes to dims from #5522, which was also marked as 0.5.0.

I'm surprised though that this bug exists after #4889 and before #5522...

**Comment by jni** [comment]
@brisvag 

> I'm surprised though that this bug exists after https://github.com/napari/napari/pull/4889 and before https://github.com/napari/napari/pull/5522...

We actually discussed this in #5522 ([link](https://github.com/napari/napari/pull/5522#discussion_r1155247092)). I said:

> I'm super curious/concerned about off by one / off by a rounding error here. I'm getting vibes of https://github.com/napari/napari/issues/1686 (and I think other similar bugs, though I can't find the other references), because generally `np.arange(rng.start, rng.stop, rng.step)` will have either `nsteps` or `nsteps+1` elements depending on floating point rounding error here. Do you have a good grip on those issues here?

To which you replied:

> I don't have a good grip... But this issue is kind of expected since we're converting from a continuous space to integers. As for the floating point precision, doing it this way (simple division) should be ok based on [previous issues](https://github.com/napari/napari/pull/4889).

😂 

The fix is to use something more like linspace rather than arange, as you noted in https://github.com/napari/napari/pull/5751/files#r1178387420. But I'm not surprised that #5751 got us either all the way there or part of the way there.

**Comment by brisvag** [comment]
Good job past me I guess xD Hopefully then #5751 is indeed the final solution. So the question is: do we feel this is important enough that we should get all the #5522 + #5751 stuff in for v0.4.18?

**Comment by manerotoni** [comment]
Thanks for the quick response. Looking forward to the update.

**Comment by GenevieveBuckley** [comment]
Closing, since the issues described have been fixed by the PRs and are now merged.

These fixes will be available in the next v0.5.0 release. Anyone who wants it sooner than that can install napari from main with `pip install git+https://github.com/napari/napari.git pyqt5`
